apps/labs/posts/numpy-thread-safe-test-suite.md

```diff
@@ -0,0 +1,379 @@
+
+---
+
+## Part Four: Modifications
+
+My first modification of the NumPy's source code could be called "baby's first commit". One thread safety issue we ran into early on was the usage of [Hypothesis](github.com/HypothesisWorks/hypothesis). NumPy was using an older version of Hypothesis which lacked some thread-safety updates that the project received, so all we had to do was bump the version!
+
+```diff
+  setuptools         ; python_version >= '3.12'
+- hypothesis==6.104.1
++ hypothesis==6.137.1
+  pytest==7.4.0
+```
+
+A small one-line change, but I remember being so excited about my first change to this massive codebase. What a rush!
+
+After this, it was time to make some more substantial changes. Most test failures could be sorted into specific categories for why they were failing. I will go in detail on why they were failing, and how I went about fixing them.
+
+### 1. Setup & Teardown
+
+The first thread safety issue I tackled was the one that required the most modifications to the test suite.
+
+#### Problem
+
+One feature of pytest is [xunit setup and teardown methods](https://docs.pytest.org/en/stable/how-to/xunit_setup.html), which was inherited from unittest. Before my internship, the way NumPy typically used this feature was to define variables that various tests would use without needing redefine the same variables over and over again. Nowadays, you may see pytest [fixtures](https://docs.pytest.org/en/stable/reference/fixtures.html#fixture) used more often instead, with the prevalent usage of xunit setup and teardown throughout NumPy's test suite being a sign of its age.
+
+```python
+# example of test using xunit setup
+class TestClass:
+  def setup_method(self):
+    self.x = 2
+
+  def test_one(self):
+    assert self.x == 2
+```
+
+Nothing wrong with old code if it still works of course! Unfortunately, this feature is currently incompatible with pytest-run-parallel and how it handles running tests under multiple threads. Depending on scope, these xunit methods run before (setup) and after (teardown) each test. Even if you don't define a teardown method, pytest has implicit default teardown implementation that it calls. This removes variables that were defined during the setup. When running tests under pytest-run-parallel, this teardown is called before all the threads in a thread pool can finish running the current test. Any test that tries to access the removed variables will fail with an `AttributeError`.
+
+_**Note:** Overall, pytest isn't very thread-safe, and so a large part of the project was figuring out what to do with thread-safety issues concerning the usage of these setup methods and thread-unsafe fixtures. However, work is currently being done to improve pytest's thread-safety, as detailed in [this issue](https://github.com/pytest-dev/pytest/issues/13768) on pytest's repo._
```

Comment:
What does "not very thread-safe" mean? It would be good to be more precise here. You don't have to explain it in detail but the "not very" language is a bit too vague.